cocos/3d/CCTerrain.h

```diff
@@ -0,0 +1,462 @@
+     **/
+    float getHeight(float x, float z, Vec3 * normal= nullptr);
+
+    /**get specified position's height mapping to the terrain,use bi-linear interpolation method
+     * @param pos the position (X,Z)
+     * @param normal the specified position's normal vector in terrain . if this argument is NULL or nullptr,Normal calculation shall be skip.
+     * @return the height value of the specified position of the terrain, if the (X,Z) position is out of the terrain bounds,it shall return 0;
+     **/
+    float getHeight(Vec2 pos, Vec3*Normal = nullptr);
+
+    /**get the normal of the specified pistion in terrain
+     * @return the normal vector of the specified position of the terrain.
+     * @note the fast normal calculation may not get precise normal vector.
+     **/
+    Vec3 getNormal(int pixelX, int pixelY);
+    /**get height from the raw height filed*/
+    float getImageHeight(int pixelX, int pixelY);
+    /**show the wireline instead of the surface,Debug Use only.
+     * @Note only support desktop platform
+     **/
+    void setDrawWire(bool boolValue);
+    /**
+     * Set threshold distance of each LOD level,must equal or gereater than the chunk size
+     * @Note when invoke initHeightMap, the LOD distance will be automatic calculated.
+     */
+    void setLODDistance(float lod1, float lod2, float lod3);
+
+    /**Switch frustum Culling Flag
+     * @Note frustum culling will remarkable improve your terrain rendering performance. 
+     */
+    void setIsEnableFrustumCull(bool boolValue);
+
+    /** set the alpha map*/
+    void setAlphaMap(cocos2d::Texture2D * newAlphaMapTexture);
+    /**set the Detail Map */
+    void setDetailMap(unsigned int index, DetailMap detailMap);
+
+    // Overrides, internal use only
+    virtual void draw(cocos2d::Renderer* renderer, const cocos2d::Mat4 &transform, uint32_t flags) override;
+    /**
+     * Ray-Terrain intersection.
+     * @return the intersection point
+     */
+    Vec3 getIntersectionPoint(const Ray & ray);
+
+    /**
+     * set the MaxDetailAmount.
+     */
+    void setMaxDetailMapAmount(int maxValue);
+
+    /**
+     * Convert a world Space position (X,Z) to terrain space position (X,Z)
+     */
+    Vec2 convertToTerrainSpace(Vec2 worldSpace);
+
+    /**
+     * reset the heightmap data.
+     */
+    void resetHeightMap(const char * heightMap);
+
+    /**
+     * get the terrain's mininal height.
+     */
+    float getMinHeight();
+
+    /**
+     * get the terrain's maximum height.
+     */
+    float getMaxHeight();
+
+    /**
+     * get the terrain's AABB(in world space)
+     */
+    AABB getAABB();
+
+    /**
+     * set the skirt height ratio
+     */
+    void setSkirtHeightRatio(float ratio);
+
+    /**
+     * get the terrain's quad tree which is also the root node.
+     */
+    QuadTree * getQuadTree();
+
+    //following methods are internal use only
+    ChunkIndices lookForIndicesLODSkrit(int selfLod, bool * result);
```

Comment:
These methods are internal used only,but the visibleness of these methods are public.